src/lib/url/url.cpp

```diff
@@ -110,116 +110,89 @@ const char *Url::GetValue(const char *aName, const char *aLastValue) const
+    const char *str;
+    long long   value;
 
-    assert(!args.Init(url));
+    VerifyOrExit((str = GetValue(aName)) != nullptr, error = OT_ERROR_NOT_FOUND);
 
-    assert(!strcmp(args.GetPath(), "/dev/ttyUSB0"));
-    assert(!strcmp(args.GetValue("baudrate"), "115200"));
-    assert(args.GetValue("not-exists") == nullptr);
-    assert(args.GetValue("last-value-wrong-position", url) == nullptr);
-    assert(args.GetValue("last-value-before-url", url - 1) == nullptr);
-    assert(args.GetValue("last-value-after-url", url + sizeof(url)) == nullptr);
+    value = strtoll(str, nullptr, 0);
+    VerifyOrExit(0 <= value && value <= UINT32_MAX, error = OT_ERROR_INVALID_ARGS);
+    aValue = static_cast<uint32_t>(value);
 
-    printf("PASS %s\r\n", __func__);
+exit:
+    return error;
 }
 
-void TestSimpleNoQueryString(void)
+otError Url::ParseUint16(const char *aName, uint16_t &aValue) const
 {
-    char         url[] = "spinel:///dev/ttyUSB0";
-    ot::Url::Url args;
+    otError  error = OT_ERROR_NONE;
+    uint32_t value;
 
-    assert(!args.Init(url));
-    assert(!strcmp(args.GetPath(), "/dev/ttyUSB0"));
-    assert(args.GetValue("last-value-wrong-position", url) == nullptr);
-    assert(args.GetValue("last-value-before-url", url - 1) == nullptr);
-    assert(args.GetValue("last-value-after-url", url + sizeof(url)) == nullptr);
+    SuccessOrExit(error = ParseUint32(aName, value));
+    VerifyOrExit(value <= UINT16_MAX, error = OT_ERROR_INVALID_ARGS);
+    aValue = static_cast<uint16_t>(value);
 
-    printf("PASS %s\r\n", __func__);
-}
-
-void TestEmptyValue(void)
-{
-    char         url[] = "spinel:///dev/ttyUSB0?rtscts&baudrate=115200&verbose&verbose&verbose";
-    ot::Url::Url args;
-    const char  *arg = nullptr;
-
-    assert(!args.Init(url));
-    assert(!strcmp(args.GetPath(), "/dev/ttyUSB0"));
-    assert((arg = args.GetValue("rtscts")) != nullptr);
-    assert(args.GetValue("rtscts", arg) == nullptr);
-    assert((arg = args.GetValue("verbose", arg)) != nullptr);
-    assert((arg = args.GetValue("verbose", arg)) != nullptr);
-    assert((arg = args.GetValue("verbose", arg)) != nullptr);
-    assert((arg = args.GetValue("verbose", arg)) == nullptr);
-
-    printf("PASS %s\r\n", __func__);
+exit:
+    return error;
 }
 
-void TestMultipleProtocols(void)
+otError Url::ParseUint8(const char *aName, uint8_t &aValue) const
 {
-    char         url[] = "spinel+spi:///dev/ttyUSB0?baudrate=115200";
-    ot::Url::Url args;
+    otError  error = OT_ERROR_NONE;
+    uint32_t value;
 
-    assert(!args.Init(url));
-    assert(!strcmp(args.GetPath(), "/dev/ttyUSB0"));
-    assert(!strcmp(args.GetValue("baudrate"), "115200"));
+    SuccessOrExit(error = ParseUint32(aName, value));
+    VerifyOrExit(value <= UINT8_MAX, error = OT_ERROR_INVALID_ARGS);
+    aValue = static_cast<uint8_t>(value);
 
-    printf("PASS %s\r\n", __func__);
+exit:
+    return error;
 }
 
-void TestMultipleProtocolsAndDuplicateParameters(void)
```

Comment:
These new APIs can only support single value parameters. The parser actually supports array, such as `forkpty-arg`.

Reply:
The new added functions are designed for the single value parameters. I have moved all url related test cases to `./tests/unit/test_url.cpp`.